lib/frodata/concerns/api.rb

```diff
@@ -0,0 +1,262 @@
+      #
+      # Returns true if the entity was successfully deleted.
+      # Returns false if an error is returned from Dynamics.
+      def destroy(*args)
+        destroy!(*args)
+      rescue *exceptions
+        false
+      end
+
+      # Public: Delete a record.
+      #
+      # entity_set - The set the entity belongs to
+      # id      - The Dynamics primary key ID of the record.
+      #
+      # Examples
+      #
+      #   # Delete the lead with id  "073ca9c8-2a41-e911-a81d-000d3a1d5a0b"
+      #   client.destroy!('leads',  "073ca9c8-2a41-e911-a81d-000d3a1d5a0b")
+      #
+      # Returns true of the entity was successfully deleted.
+      # Raises an exception if an error is returned from Dynamics.
+      def destroy!(entity_set, id)
+        query = service[entity_set].query
+        url_chunk = query.find(id).to_s
+        api_delete url_chunk
+        true
+      end
+
+      # Public: Finds a single record and returns all fields.
+      #
+      # entity_set - The set the entity belongs to
+      # id      - The id of the record. If field is specified, id should be the id
+      #           of the external field.
+      # field   - External ID field to use (default: nil).
+      #
+      # Returns the Enitity record.
+      def find(entity_set, id, field = nil)
+        query = service[entity_set].query
+        url_chunk = if field
+                      query.where("#{field} eq #{id}")
+                    else
+                      query.find(id).to_s
+                    end
+        body = api_get(url_chunk).body
+        build_entity(entity_set, body)
+      end
+
+      # Public: Finds a single record and returns select fields.
+      #
+      # entity_set - The set the entity belongs to
+      # id      - The id of the record. If field is specified, id should be the id
+      #           of the external field.
+      # select  - A String array denoting the fields to select.  If nil or empty array
+      #           is passed, all fields are selected.
+      # field   - External ID field to use (default: nil).
+      #
+      def select(entity_set, id, select, field = nil)
+        query = service[entity_set].query
+
+        select.each{|field| query.select(field)}
+
+        url_chunk = if field
+                      query.where("#{field} eq #{id}")
+                    else
+                      query.find(id).to_s
+                    end
+
+        body = api_get(url_chunk).body
+        build_entity(entity_set, body)
+      end
+
+      private
+
+      # Internal: Returns a path to an api endpoint
+      #
+      # Examples
+      #
+      #   api_path('leads')
+      #   # => '/api/data/v9.1/leads'
+      def api_path(path)
+        "/api/data/v#{options[:api_version]}/#{path}"
+      end
+
+      def build_entity(entity_set, data)
+        entity_options = service[entity_set].entity_options
+        single_entity?(data) ? parse_entity(data, entity_options) : parse_entities(data, entity_options)
```

Comment:
Instead of having build_entity, I would have build_entity and build_entities. Based on the query you already know which one type of result will be returned. There isn't a need to guess.

Also if its a searching by key and doesn't find a value I think it returns a 404. Versus if you are querying for multiple values it will return and empty list

Reply:
404 will raise an error

there is no really guessing but was making a simple abstraction to avoid repetition and it is easy to figure. I ll leave it as is for now